src/platform/qt/library/LibraryController.h

```diff
@@ -0,0 +1,135 @@
+/* Copyright (c) 2014-2017 waddlesplash
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+#pragma once
+
+#include <QList>
+#include <QMap>
+#include <QSharedPointer>
+#include <QStackedWidget>
+
+#include <mgba/core/library.h>
+
+namespace QGBA {
+
+// Predefinitions
+class LibraryGrid;
+class LibraryTree;
+class ConfigController;
+
+enum class LibraryStyle {
+	STYLE_LIST,
+	STYLE_TREE,
+	STYLE_GRID,
+	STYLE_ICON
+};
+
+static QString NiceSizeFormat(size_t filesize)
+{
+	double size = filesize;
+	QString unit = "B";
+	if (size >= 1024.0) {
+		size /= 1024.0;
+		unit = "kiB";
+	}
+	if (size >= 1024.0) {
+		size /= 1024.0;
+		unit = "MiB";
+	}
+	return QString("%0 %1").arg(size, 0, 'f', 1).arg(unit);
+}
+static QString NicePlatformFormat(mPlatform platform)
+{
+	switch (platform) {
+#ifdef M_CORE_GBA
+	case PLATFORM_GBA:
+		return QObject::tr("GBA");
+#endif
+#ifdef M_CORE_GB
+	case PLATFORM_GB:
+		return QObject::tr("GB");
+#endif
+	default:
+		return QObject::tr("?");
+	}
+}
+
+class CLibraryEntry {
+public:
+	CLibraryEntry(mLibraryEntry* entry);
+	~CLibraryEntry();
+
+	QString displayTitle() const { return title().isNull() ? filename() : title(); }
+
+	QString base() const { return QString(entry->base); }
+	QString filename() const { return QString(entry->filename); }
+	QString fullpath() const { return m_fullpath; }
+	QString title() const { return QString(entry->title); }
+	QByteArray internalTitle() const { return QByteArray(entry->internalTitle); }
+	QByteArray internalCode() const { return QByteArray(entry->internalCode); }
+	mPlatform platform() const { return entry->platform; }
+	size_t filesize() const { return entry->filesize; }
+	uint32_t crc32() const { return entry->crc32; }
+
+private:
+	friend class LibraryController;
+	mLibraryEntry* entry;
+	const QString m_fullpath;
+};
+typedef QSharedPointer<CLibraryEntry> LibraryEntry;
+
+class AbstractGameList {
+public:
+	virtual LibraryEntry selectedEntry() = 0;
+	virtual void selectEntry(LibraryEntry game) = 0;
+
+	virtual void setViewStyle(LibraryStyle newStyle) = 0;
+
+	virtual void addEntry(LibraryEntry item) = 0;
+	virtual void addEntries(QList<LibraryEntry> items);
+
+	virtual void removeEntry(LibraryEntry item) = 0;
+	virtual void removeEntries(QList<LibraryEntry> items);
+};
+
+class LibraryController final : public QStackedWidget {
+Q_OBJECT
+
+public:
+	LibraryController(QWidget* parent_widget = nullptr, const QString& path = "",
```

Comment:
Huh? Zero? Basically all QWidget-derived classes have it as the first argument. I can't think of any that don't. If you reverse the order, the compile fails because the uic-generated files expect it to be the first argument.